### PR TITLE
BUILD-87: secret configmap volume mounts in builds

### PR DESCRIPTION
**Pull request by coreydaley**
Adding fields to the build api to support mounting Secrets and ConfigMaps within build containers as volumes.

https://issues.redhat.com/browse/BUILD-87

### CODE REVIEW

File: build/v1/types.go
Comment (gabemontero) [inline build/v1/types.go:747]: Would it make sense to have an encapsulating struct for the two arrays here?  I think it makes sense but will acquiesce to you two @adambkaplan @coreydaley if you all disagree.

My reasoning:  This is a bit different in the k8s case where volumes are defined at the pod spec level but volume mounts are at the container level.

For use, we have stuff at the same level, and certainly we don't envisioning using 1 array without the other.

**Review by gabemontero — COMMENTED** [review]
just one suggestion @coreydaley @adambkaplan

**Review by adambkaplan — COMMENTED** [review]
API generally looks good, just a few nits on docs to eventually consider.

We shouldn't merge this until we are confident this will land in 4.8

File: build/v1/types.go
Comment (adambkaplan) [inline build/v1/types.go:1410]: @coreydaley In reviewing the apiserver implementation, we need the following:

1. Mark the field optional
2. Add the k8s "no-generate" marker.
3. Cast the protobuf type

```suggestion
        // +k8s:conversion-gen=false
        // +optional
	Type BuildVolumeSourceType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=BuildVolumeSourceType"`
```

File: build/v1/types.go
Comment (adambkaplan) [reply build/v1/types.go:747]: Makes sense - so our API could look like this:

```yaml
volumes:
- name: npm-secret
  secret:
    name: npm-secret
  mount:
    destinationPath: /var/run/secrets/npm-secret
    readOnly: true
```

@bparees thoughts?

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1386]: updated in next push

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1410]: updated in next push

**Comment by coreydaley** [comment]
@adambkaplan updated per your comments, ptal

File: build/v1/types.go
Comment (bparees) [reply build/v1/types.go:747]: it does seem to remove a potential source of error (mismatched volume/volumemount names).

the only downside i can think of offhand is that you can't do something like mount a single volume in two different locations if you wanted to (without explicitly defining the volume twice also).  But i can't see that being critical.

so i think i'm +1 on @gabemontero's suggestion.

Comment (coreydaley) [reply build/v1/types.go:747]: How about:

```
volumes:
- name: npm-secret
  source:
	type: [secret/configmap]
  	secret:
    	    name: npm-secret
  mount:
    destinationPath: /var/run/secrets/npm-secret
    readOnly: true
```

Comment (adambkaplan) [reply build/v1/types.go:747]: > the only downside i can think of offhand is that you can't do something like mount a single volume in two different locations if you wanted to (without explicitly defining the volume twice also). But i can't see that being critical.

Could this become an issue down the road if someone wants to mount a PVC in two locations? I think yes.

So to take @coreydaley 's suggestion (which I like), we can change `mount` to a `mounts` array, keyed off of the destination path:

```yaml
volumes:
- name: npm-secret
  source:
    type: [secret/configmap]
    secret:
      name: npm-secret
  mounts:
  -  destinationPath: /var/run/secrets/npm-secret
     readOnly: true
```

**Comment by coreydaley** [comment]
/retitle BUILD-87 secret configmap volume mounts in builds

**Comment by coreydaley** [comment]
@gabemontero @adambkaplan @bparees Updated per your comments, ptal

File: build/v1/types.go
Comment (adambkaplan) [inline build/v1/types.go:747]: We also need to set a patch merge strategy so that volumes can be appended with `kubectl apply`/`kubectl patch`:

```suggestion
	Volumes []BuildVolume `json:"volumes,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,9,opt,name=volumes"`
```

Comment (coreydaley) [reply build/v1/types.go:747]: updated in next push

File: build/v1/types.go
Comment (adambkaplan) [inline build/v1/types.go:1430]: More of a nit - since `DestinationPath` is required, it should be listed first in the struct and protobuf.

```suggestion
	// destinationPath is the path within the build runtime environment at which the volume should be mounted.
	// Must not contain ':' and can not collide with a destination path generated by the builder process.
	DestinationPath string `json:"destinationPath" protobuf:"bytes,1,opt,name=destinationPath"`
	
	// readOnly denotes whether the volume should be mounted as readonly or not
	// The volume is mounted read-only if true, read-write otherwise (false or unspecified).
	// Defaults to false.
	// +optional
	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,2,opt,name=readOnly"`
```

Comment (coreydaley) [reply build/v1/types.go]: updated in current push

**Comment by coreydaley** [comment]
@adambkaplan @bparees Since VolumeMounts in a container have to have the same name is the Volume that they are mounting from the pod, can two VolumeMounts have the same name? If not, having `Mounts []BuildVolumeMount` won't work, since those inherit the names from the BuildVolume.  Thoughts?

**Comment by bparees** [comment]
> Since VolumeMounts in a container have to have the same name is the Volume that they are mounting from the pod, 
can two VolumeMounts have the same name? If not, having Mounts []BuildVolumeMount won't work, since those inherit the names from the BuildVolume. Thoughts?

if someone wants to mount the same content into two different locations, you'd still only mount it into the build pod once.

you'd then mount it into the buildah container twice, but then you aren't naming anything, you're just passing paths/mount locations. (I am assuming buildah allows you to do something like "--volume /pod/filesystem:/buildah/path1 --volume /pod/filesytem:/buildah/path2", but if buildah chokes on refing the same source path in two mounts, then that's another issue)

**Comment by coreydaley** [comment]
> > Since VolumeMounts in a container have to have the same name is the Volume that they are mounting from the pod,
> > can two VolumeMounts have the same name? If not, having Mounts []BuildVolumeMount won't work, since those inherit the names from the BuildVolume. Thoughts?
> 
> if someone wants to mount the same content into two different locations, you'd still only mount it into the build pod once.
> 
> you'd then mount it into the buildah container twice, but then you aren't naming anything, you're just passing paths/mount locations. (I am assuming buildah allows you to do something like "--volume /pod/filesystem:/buildah/path1 --volume /pod/filesytem:/buildah/path2", but if buildah chokes on refing the same source path in two mounts, then that's another issue)

A `corev1.VolumeMount`, which gets mounted into the `pod.Spec.Containers[0]` has a `Name` field, which must match the `Volume.Name` correct? So if there are two VolumeMounts for one Volume ...

**Comment by adambkaplan** [comment]
> A `corev1.VolumeMount`, which gets mounted into the `pod.Spec.Containers[0]` has a `Name` field, which must match the `Volume.Name` correct? So if there are two VolumeMounts for one Volume ...

This confusion is another reason why having an API with a slightly different look and feel than k8s Volume/VolumeMount API is a good thing. Having worked through the feature design, I now use the term _buildah runtime environment_ to describe where buildah runs its build. I try (with various degrees of success) to use the term "build pod" and "build container" to refer to the kubernetes pod/container that then invokes buildah.

What the API describes is how we get a Kubernetes volume to a volume mount for the _buildah runtime environment_. This API does _not_ describe how we get the Kubernetes volume mounted into the build pod/build container - that is an implementation detail within the build controller and the openshift/builder binary.

For the build pod/container, we only need a 1:1 mapping between the kubernetes volume and the container volume mount. As @bparees mentioned, when we set up buildah's runtime environment we can mount the same source directory in more than one destination.

File: build/v1/types.go
Comment (adambkaplan) [inline build/v1/types.go:1430]: This implies that when we set up the volume mount in the build pod containers, we either:

1. Always mount as read-write.
2. Be smart about setting read-only. For example, set the build container volume mount as read-only if and only if all buildah volume mounts are read only.

@nalind can we run into issues if the same directory is mounted into buildah in two different locations with different read/write settings?

@jsafrane must we always mount Secrets and ConfigMaps into containers as read-only? Is mounting a Secret or ConfigMap as a writeable directory impossible, or merely a bad idea?

Comment (coreydaley) [reply build/v1/types.go:1430]: @bparees So, during our discussion I thought that we clarified that the volumeMounts don't need to be setup in the container? Only the buildah mounts?

Comment (bparees) [reply build/v1/types.go:1430]: we clarified that you have total freedom to decide how to get the content into the build pod/container (using adam's terminology).  It will probably be *a* volume/volumeount, but it doesn't have to be the volume/volumemount as specified by the user (in terms of the volume name and the volumemount path).

Comment (bparees) [reply build/v1/types.go:1430]: as regards read/readwrite...  for secrets/configmaps it really doesn't matter (nothing you do is going to change the actual secret/configmap resource).  If/when this is extended to cover persistent volumes for caching build data, how you mount the volume into the build pod/container will matter, however.

**Comment by adambkaplan** [comment]
/hold

Given the complexity of the feature and the time remaining before 4.8 reaches "final freeze" period, we are going to hold this until master opens for 4.9

File: build/v1/types.go
Comment (jsafrane) [reply build/v1/types.go:1409]: > must we always mount Secrets and ConfigMaps into containers as read-only?

Yes. See here: https://github.com/kubernetes/kubernetes/issues/62099 (+ linked PR)

Comment (adambkaplan) [reply build/v1/types.go:1409]: @bparees @coreydaley thoughts on dropping this field, and later introducing a field like `Writeable` for read-write volumes?

The current implementation has us do the following:

1. Mount the volume into the builder container in a fixed directory - ex `/var/run/build/volumes/xxxx`
2. Tell buildah to use the directory above as a transient bind mount, passing the `ro` or `rw` option as specified by the user.

If a Secret or ConfigMap must always be mounted as a read only filesystem, then having the user bind mount said read-only filesystem in buildah as a writeable directory will lead to inevitable errors. To avoid these errors, we would need to copy the contents to a writeable directory, which feels like overkill.

Comment (coreydaley) [reply build/v1/types.go:1409]: Dropping the field for now makes sense to me, since it is not needed.

Comment (coreydaley) [reply build/v1/types.go]: @adambkaplan updated in current push, ptal

**Comment by adambkaplan** [comment]
@xiuwang @rolfedh @RickJWagner coding for this feature is ready for QE review and docs/PX acks/approvals.

File: build/v1/types.go
Comment (adambkaplan) [inline build/v1/types.go:1406]: A quick note regarding this API.

`Secrets` and `ConfigMaps` require read-only volumes to work in Kubernetes. Rather than forcing users to specify `ReadOnly`, we decided to drop this field and assume volumes will be mounted read-only by default.

If in the future we support writeable volumes (ex - PVCs), the API will need to expose a field like `Writeable` as an option, with validations to ensure that builds with Secret or ConfigMap volumes to not have this value set to true.

**Review by adambkaplan — APPROVED** [review]
/lgtm

API looks good from the build team's side.

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1358]: can you specify what happens if it does collide, here?  Users of the api will not have any idea what volumes the build controller adds in order to avoid them, so even better might be providing some indication of the naming scheme the build controller uses so users can avoid colliding.

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1366]: buildah is an implementation detail.  "read/write status of the volume in the image build container" ?

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1386]: list the valid values

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1367]: did we actually confirm that a single path can be mounted inside the buildah container multiple times?

File: build/v1/types.go
Comment (bparees) [reply build/v1/types.go:1406]: so to be clear, you're mounting it as writable within the buildah container (so dockerfile commands can write to the mounted dir path) but those write operations will not propagate to the actual k8s resource backing the pod volume?

or will writes from the Dockerfile commands fail because even the buildah path is read-only?

i would guess it's the latter.

Regardless, the behavior should be specified.

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1405]: do we require this to be an absolute path?  and to not contain ".."?  (those are things that have gotten us in trouble in the past)

**Review by bparees — CHANGES_REQUESTED** [review]

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1386]: Is this ok for the list, or is there a specific format that is preferred?
`// Valid types are: Secret, ConfigMap`

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1367]: I am launching a cluster to verify and will report back.

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1358]: We actually take the `name` that they provide and append `-user-build-volume` to it, so there is probably a very very low chance of collision. I also don't believe that the build controller has a naming scheme that it uses for other types of mounts. They just are what they are. I can provide an example once my cluster finished launching

File: build/v1/types.go
Comment (bparees) [reply build/v1/types.go:1386]: that looks fine to me

File: build/v1/types.go
Comment (bparees) [reply build/v1/types.go:1358]: it does have a naming scheme, though there may be several for different types of things being mounted.  But for example for secrets i think it uses a pattern like "Sourcesecret_1", "sourcesecret_2" or some such thing.

Comment (bparees) [reply build/v1/types.go:1358]: anyway the main clarification i want to see is rather than saying "it cannot collide" it should either say:
1) names that collide will result in a validation error
or
2) whatever the behavior is if a collision occurs

Comment (coreydaley) [reply build/v1/types.go:1358]: updated in next push

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1405]: Right now we do not do any validation on the path. @adambkaplan thoughts?

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1367]: Yes, a single path can be mounted inside the buildah container multiple times:
```
volumes:
      - mounts:
        - destinationPath: /var/run/secrets/some-secret
        - destinationPath: /var/run/secrets/some-secret-2
        name: this-secret
        source:
          secret:
            defaultMode: 420
            secretName: my-secret
          type: Secret
```
```
STEP 11: RUN /bin/sh -ic 'cat /var/run/secrets/some-secret/key; cat /var/run/secrets/some-secret-2/key'
my-secret-value my-secret-value
```

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1406]: The volumeMounts within the containers pod definition are readOnly:
```
- mountPath: /var/run/openshift.io/volumes/my-secret-user-build-volume
      name: my-secret-user-build-volume
      readOnly: true
    - mountPath: /var/run/openshift.io/volumes/my-configmap-user-build-volume
      name: my-configmap-user-build-volume
      readOnly: true
```
And the transient mount definition that is sent to buildah is readonly also:
```
I0614 14:51:34.597118       1 daemonless.go:286] transient mounts: []string{"/tmp/rhsm-copy257734809:/run/secrets/rhsm:nodev,noexec,nosuid", "/var/run/openshift.io/volumes/my-secret-user-build-volume:/var/run/secrets/some-secret:ro", "/var/run/openshift.io/volumes/my-secret-user-build-volume:/var/run/secrets/some-secret-2:ro", "/var/run/openshift.io/volumes/my-configmap-user-build-volume:/var/run/configmaps/some-configmap:ro"}
```

File: build/v1/types.go
Comment (adambkaplan) [reply build/v1/types.go:1405]: I think both of these make sense - we can add this in openshift-apiserver.

Comment (coreydaley) [reply build/v1/types.go:1405]: @bparees @adambkaplan added validations and tests for validations (100% coverage) as per your comments: https://github.com/openshift/openshift-apiserver/pull/208/commits/849c39674b40d378cb423a4a65d1945a17245893

**Comment by xiuwang** [comment]
Test on PR launched cluster `launch openshift/builder#245,openshift/openshift-controller-manager#183,openshift/openshift-apiserver#208,openshift/oc#843`  following test cases  https://url.corp.redhat.com/buildvolumes , no critical issues found.

/label qe-approved

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go]: addressed in next push

**Comment by coreydaley** [comment]
@bparees @adambkaplan ptal, all comments have been addressed.

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:746]: don't mention buildah, it's implementation specific and "buildah's runtime environment" doesn't mean much to a user.  Explain it in terms of what it means for their consumption of the content.

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:783]: same here

File: build/v1/types.go
Comment (bparees) [inline build/v1/types.go:1386]: what's the point in even having this if we're going to infer it?

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:746]: updated in next push

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:783]: updated in next push

File: build/v1/types.go
Comment (coreydaley) [reply build/v1/types.go:1386]: I'm not sure, @adambkaplan might have a better idea since he wrote the ep.

Comment (adambkaplan) [reply build/v1/types.go:1386]: This object is an example of a Union object, as describe by the upstream api conventions [1].
Scanning the various types of union objects upstream, it seems that some, but not all, objects employ the _type discriminator_ pattern, which is what I've advocated for here.
For instance, in the `Deployment` object, a union with a type discriminator is used for the deployment strategy [2].
However, type discriminator usage is not universal, most notably with the VolumeSource object.
It isn't clear if this is intentional or evidence of tech debt that can't be paid off.

My understanding, based on discussions with @deads2k, is that type discriminators with union objects help controllers in version skew situations. As an example, we can introduce a `PVC` volume type in v4.10. During an upgrade, we could have a cluster state where a 4.10 apiserver allows PVC volumes for builds, but the controller remains on 4.9 and can't start the build. The `Type` field lets us give a cleaner error message ("volume type PVC is not supported") versus a more vague and perhaps unfriendly error message (ex: "no volume source information was specified").

@deads2k @sttts I don't see any any recommendations from sig-architecture regarding type discriminator fields. Is this pattern something that is recommended upstream, something that we emphasize for OpenShift, or neither (i.e. this practice is outdated and we shouldn't do it)?

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions
[2] https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/deployment-v1/#DeploymentSpec

Comment (bparees) [reply build/v1/types.go:1386]: i'm less concerned with having it or not having and more concerned w/ it being optional.

either have it and make users specify it, or don't have it and handle the migration challenges in code, would be my expectation.

having it but having it be optional means it's not providing anything you can't already determine by looking at which other fields are populated.

Comment (deads2k) [reply build/v1/types.go:1386]: Making the Type required makes sense for this API.  We allow it to be optional in some cases in our configuration APIs and ascribe a "choose a good default for me which may change over time".  In this case, there is no reasonable default which can be chosen: additional information is always required for a functional API.

Both the error message and the clear intent are valuable for validation and client skew.  Using a Type field for unions is recommended.

Comment (coreydaley) [reply build/v1/types.go:1386]: updated in next push.

**Comment by coreydaley** [comment]
@bparees @adambkaplan ptal
I believe that all comments have been addressed and the Type field has been changed to required and the comment updated.

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Pull request merged**
